components/layout_2020/fragments.rs

```diff
@@ -375,8 +385,45 @@ impl Serialize for TextFragment {
 
 impl Serialize for ImageFragment {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
-        let mut serializer = serializer.serialize_struct("ImageFragment", 1)?;
+        let mut serializer = serializer.serialize_struct("ImageFragment", 2)?;
+        serializer.serialize_field("debug_id", &self.debug_id)?;
         serializer.serialize_field("rect", &self.rect)?;
         serializer.end()
     }
 }
+
+#[cfg(not(debug_assertions))]
+#[derive(Clone)]
+pub struct DebugId;
+
+#[cfg(debug_assertions)]
+#[derive(Clone)]
+pub struct DebugId(u16);
+
+#[cfg(not(debug_assertions))]
+impl DebugId {
+    pub fn new() -> DebugId {
+        DebugId
+    }
+}
+
+#[cfg(debug_assertions)]
+impl DebugId {
+    pub fn new() -> DebugId {
+        DebugId(layout_debug::generate_unique_debug_id())
+    }
+}
+
+#[cfg(not(debug_assertions))]
+impl Serialize for DebugId {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        serializer.serialize_str(&format!("{:p}", &self))
+    }
+}
+
+#[cfg(debug_assertions)]
+impl Serialize for DebugId {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        serializer.serialize_u16(self.0)
```

Comment:
This impl can be derived with `#[serde(transparent)]`

Reply:
I just learnt about this. Thanks!